node/src/main/scala/com/wavesplatform/state/BlockchainUpdated.scala

```diff
@@ -0,0 +1,89 @@
+package com.wavesplatform.state
+
+import cats.implicits._
+import com.wavesplatform.account.Address
+import com.wavesplatform.block.{Block, MicroBlock}
+import com.wavesplatform.common.state.ByteStr
+import com.wavesplatform.state.diffs.BlockDiffer.DetailedDiff
+import com.wavesplatform.transaction.Asset
+import com.wavesplatform.transaction.Asset.Waves
+import com.wavesplatform.state.reader.CompositeBlockchain.composite
+import monix.reactive.Observer
+
+final case class StateUpdated(balances: Seq[(Address, Asset, Long)], leases: Seq[(Address, LeaseBalance)], dataEntries: Seq[(Address, DataEntry[_])])
+
+sealed trait BlockchainUpdated
+final case class BlockAdded(block: Block, height: Int, blockStateUpdate: StateUpdated, transactionsStateUpdates: Seq[StateUpdated])
+    extends BlockchainUpdated
+final case class MicroBlockAdded(microBlock: MicroBlock,
+                                 height: Int,
+                                 microBlockStateUpdate: StateUpdated,
+                                 transactionsStateUpdates: Seq[StateUpdated])
+    extends BlockchainUpdated
+final case class RollbackCompleted(to: ByteStr, height: Int)           extends BlockchainUpdated
+final case class MicroBlockRollbackCompleted(to: ByteStr, height: Int) extends BlockchainUpdated
+
+object BlockchainUpdateNotifier {
+
+  private def stateUpdateFromDiff(blockchain: Blockchain, diff: Diff): StateUpdated = {
```

Comment:
There's already a method which does precisely that: see `Caches.append`. To avoid duplicating the logic, consider extracting common pieces into shared methods.